File: agents/learning.ts
import { BaseAgent } from './base';
import type { AgentMetrics, EvaluationResult } from '../types';

export class LearningAgent extends BaseAgent {
  constructor(
    model: ReturnType<typeof import('../utils/glm-client').getGLMConfig>,
    tools: ReturnType<typeof import('../tools/registry').ToolRegistry>,
    kb: ReturnType<typeof import('../learning/knowledge-base').KnowledgeBase>,
  ) {
    const systemPrompt = `You are a Learning Agent focused on system improvement.

Your responsibilities:
1. Analyze execution results to identify patterns
2. Evaluate code quality and effectiveness
3. Identify successful vs unsuccessful approaches
4. Update knowledge base with learned patterns
5. Suggest improvements to system behavior
6. Monitor performance metrics

When analyzing:
- Look for code patterns that consistently succeed
- Identify common failure modes
- Suggest optimizations
- Be specific and actionable`;

    super(model, tools, kb, systemPrompt);
    this.metrics: AgentMetrics = {
      totalTasks: 0,
      successfulTasks: 0,
      patternsLearned: 0,
    };
  }

  async evaluateTask(taskResult: TaskResult): Promise<EvaluationResult> {
    this.metrics.totalTasks += 1;

    if (taskResult.success) {
      this.metrics.successfulTasks += 1;
    }

    const prompt = `Evaluate this task execution:

Task: ${taskResult.task.description}
Success: ${taskResult.success}
Output: ${taskResult.output.substring(0, 500)}

Provide assessment:
1. Success/Failure analysis
2. What went well (if successful)
3. What went wrong (if failed)
4. Key learnings
5. Recommendations for improvement`;

    const evaluation = await this.think(prompt, false);

    return {
      evaluation,
      success: taskResult.success,
      metrics: { ...this.metrics },
    };
  }

  async extractPattern(taskData: TaskResult): Promise<any> {
    if (!taskData.success) {
      return null;
    }

    const prompt = `Extract reusable patterns from this successful task:

Description: ${taskData.task.description}
Output: ${taskData.output.substring(0, 1000)}

Identify:
1. Code patterns that could be reused
2. Approaches that worked well
3. General principles that apply

Format as JSON with 'patterns' array containing:
- type: pattern type (e.g., api_endpoint, data_model, test_pattern)
- code: pattern code
- description: when and why to use this pattern
- complexity: low/medium/high`;

    const response = await this.think(prompt, false);

    try {
      const patternData = JSON.parse(response);
      if (patternData.patterns) {
        this.metrics.patternsLearned += patternData.patterns.length;
      }
      return patternData;
    } catch (error) {
      return null;
    }
  }

  async suggestImprovements(performanceData: any): Promise<string[]> {
    const successRate =
      this.metrics.totalTasks > 0
        ? this.metrics.successfulTasks / this.metrics.totalTasks
        : 0;

    const prompt = `Analyze system performance and suggest improvements:

Total Tasks: ${this.metrics.totalTasks}
Success Rate: ${(successRate * 100).toFixed(1)}%
Patterns Learned: ${this.metrics.patternsLearned}

Recent Performance:
${JSON.stringify(performanceData).substring(0, 1000)}

Suggest 3-5 concrete improvements to:
1. Increase success rate
2. Better utilize knowledge base
3. Improve code quality
4. Speed up execution

Format as a numbered list.`;

    const suggestionsText = await this.think(prompt, false);

    const suggestions = [];
    for (const line of suggestionsText.split('\n')) {
      if (line.trim() && line.trim()[0].match(/\d/)) {
        suggestions.push(line.trim());
      }
    }

    return suggestions;
  }

  async generateReport(): Promise<string> {
    const successRate =
      this.metrics.totalTasks > 0
        ? this.metrics.successfulTasks / this.metrics.totalTasks
        : 0;

    const report = `
Learning Agent Report
==================

Performance Metrics:
- Total Tasks: ${this.metrics.totalTasks}
- Successful Tasks: ${this.metrics.successfulTasks}
- Success Rate: ${(successRate * 100).toFixed(1)}%
- Patterns Learned: ${this.metrics.patternsLearned}

System Health: ${successRate > 0.7 ? 'Healthy' : 'Needs Improvement'}

Top Recommendations:
`;

    const recommendations = await this.suggestImprovements({});
    report += recommendations.slice(0, 5).map((r, i) => `- ${r}`).join('\n');

    return report;
  }

  async improveFromFeedback(feedback: string): Promise<any> {
    const prompt = `Analyze and incorporate this user feedback:

${feedback}

Update system behavior and knowledge base to address feedback.
Provide:
1. What was learned
2. How system will change
3. Knowledge base updates needed`;

    const improvements = await this.think(prompt, false);

    return {
      feedback,
      improvements,
      applied: true,
    };
  }
}

File: glm_code_system/agents/learning.py
"""Learning agent for evaluation and continuous improvement."""

from typing import Any

from glm_code_system.agents.base import BaseAgent


class LearningAgent(BaseAgent):
    """Agent responsible for learning and system improvement."""

    def __init__(
        self,
        model: Any,
        tools: Any,
        knowledge_base: Any,
    ) -> None:
        """Initialize learning agent."""
        system_prompt = """You are a Learning Agent focused on system improvement.

Your responsibilities:
1. Analyze execution results to identify patterns
2. Evaluate code quality and effectiveness
3. Identify successful vs unsuccessful approaches
4. Update knowledge base with learned patterns
5. Suggest improvements to system behavior
6. Monitor performance metrics

When analyzing:
- Look for code patterns that consistently succeed
- Identify common failure modes
- Suggest optimizations
- Be specific and actionable"""
        super().__init__(model, tools, knowledge_base, system_prompt)
        self.metrics: dict[str, Any] = {
            "total_tasks": 0,
            "successful_tasks": 0,
            "patterns_learned": 0,
        }

    async def evaluate_task(
        self,
        task_result: dict[str, Any],
    ) -> dict[str, Any]:
        """Evaluate the outcome of a task."""
        self.metrics["total_tasks"] += 1

        if task_result.get("success", False):
            self.metrics["successful_tasks"] += 1

        prompt = f"""Evaluate this task execution:

Task: {task_result.get('task', {}).get('description', 'Unknown')}
Success: {task_result.get('success', False)}
Output: {task_result.get('output', '')[:500]}

Provide assessment:
1. Success/Failure analysis
2. What went well (if successful)
3. What went wrong (if failed)
4. Key learnings
5. Recommendations for improvement"""

        evaluation = await self.think(prompt, use_memory=False)

        return {
            "evaluation": evaluation,
            "success": task_result.get("success", False),
            "metrics": self.metrics.copy(),
        }

    async def extract_pattern(
        self,
        task_data: dict[str, Any],
    ) -> dict[str, Any] | None:
        """Extract reusable patterns from task execution."""
        if not task_data.get("success"):
            return None

        prompt = f"""Extract reusable patterns from this successful task:

Description: {task_data.get('task', {}).get('description', '')}
Output: {task_data.get('output', '')[:1000]}

Identify:
1. Code patterns that could be reused
2. Approaches that worked well
3. General principles that apply

Format as JSON with 'patterns' array containing:
- type: pattern type (e.g., api_endpoint, data_model, test_pattern)
- code: the pattern code
- description: when and why to use this pattern
- complexity: low/medium/high"""

        response = await self.think(prompt, use_memory=False)

        try:
            import json

            pattern_data = json.loads(response)
            if "patterns" in pattern_data:
                self.metrics["patterns_learned"] += len(pattern_data["patterns"])
                return pattern_data
        except (json.JSONDecodeError, KeyError):
            pass

        return None

    async def suggest_improvements(
        self,
        performance_data: dict[str, Any],
    ) -> list[str]:
        """Suggest system improvements based on performance."""
        success_rate = 0
        if self.metrics["total_tasks"] > 0:
            success_rate = (
                self.metrics["successful_tasks"] / self.metrics["total_tasks"]
            )

        prompt = f"""Analyze system performance and suggest improvements:

Total Tasks: {self.metrics['total_tasks']}
Success Rate: {success_rate:.1%}
Patterns Learned: {self.metrics['patterns_learned']}

Recent Performance:
{json.dumps(performance_data, indent=2)[:1000]}

Suggest 3-5 concrete improvements to:
1. Increase success rate
2. Better utilize knowledge base
3. Improve code quality
4. Speed up execution

Format as a numbered list."""

        suggestions_text = await self.think(prompt, use_memory=False)

        suggestions = []
        for line in suggestions_text.split("\n"):
            line = line.strip()
            if line and line[0].isdigit():
                suggestions.append(line)

        return suggestions

    async def generate_report(self) -> str:
        """Generate learning report."""
        success_rate = 0
        if self.metrics["total_tasks"] > 0:
            success_rate = (
                self.metrics["successful_tasks"] / self.metrics["total_tasks"]
            )

        report = f"""
Learning Agent Report
==================

Performance Metrics:
- Total Tasks: {self.metrics['total_tasks']}
- Successful Tasks: {self.metrics['successful_tasks']}
- Success Rate: {success_rate:.1%}
- Patterns Learned: {self.metrics['patterns_learned']}

System Health: {'Healthy' if success_rate > 0.7 else 'Needs Improvement'}

Top Recommendations:
"""
        recommendations = await self.suggest_improvements({})
        report += "\n".join([f"- {r}" for r in recommendations[:5]])

        return report

    async def improve_from_feedback(
        self,
        feedback: str,
    ) -> dict[str, Any]:
        """Incorporate user feedback into learning."""
        prompt = f"""Analyze and incorporate this user feedback:

{feedback}

Update system behavior and knowledge base to address feedback.
Provide:
1. What was learned
2. How system will change
3. Knowledge base updates needed"""

        improvements = await self.think(prompt)

        return {
            "feedback": feedback,
            "improvements": improvements,
            "applied": True,
        }
